File: git-touch/git-touch.php
<?php
declare(strict_types = 1);

$r = popen('git ls-files', 'r');
$file_n = 0;

while (true) {
   $get_s = fgets($r);
   if (feof($r)) {
      break;
   }
   $trim_s = rtrim($get_s);
   $file_m[$trim_s] = false;
   $file_n++;
}

$r = popen('git log -m -z --name-only --relative --format=%n%ct .', 'r');
fgets($r);

while ($file_n > 0) {
   $unix = (int)rtrim(fgets($r));
   $names = explode("\x0", rtrim(fgets($r)));
   foreach ($names as $name) {
      if (! key_exists($name, $file_m)) {
         continue;
      }
      if ($file_m[$name]) {
         continue;
      }
      echo $unix, "\t", $name, "\n";
      touch($name, $unix);
      $file_m[$name] = true;
      $file_n--;
   }
}

File: git-touch/git-touch.go
package main

import (
   "github.com/89z/x"
   "log"
   "os"
   "strconv"
   "time"
)

func main() {
   gitLs, e := x.Popen("git", "ls-files")
   if e != nil {
      log.Fatal(e)
   }
   files := map[string]bool{}
   for gitLs.Scan() {
      files[gitLs.Text()] = true
   }
   gitLog, e := x.Popen(
      "git", "log", "-m",
      "--name-only", "--relative", "--pretty=format:%ct", ".",
   )
   if e != nil {
      log.Fatal(e)
   }
   for len(files) > 0 {
      gitLog.Scan()
      sec, e := strconv.ParseInt(gitLog.Text(), 10, 64)
      if e != nil {
         log.Fatal(e)
      }
      unix := time.Unix(sec, 0)
      for gitLog.Scan() {
         name := gitLog.Text()
         if name == "" {
            break
         }
         if ! files[name] {
            continue
         }
         println(sec, "\t", name)
         os.Chtimes(name, unix, unix)
         delete(files, name)
      }
   }
}
